src/lightning_lite/plugins/collectives/collective.py

```diff
@@ -0,0 +1,129 @@
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+import torch
+from typing_extensions import Protocol, runtime_checkable, Self
+
+
+@runtime_checkable
+class CollectibleGroup(Protocol):
+    def size(self) -> int:
+        ...
+
+    def rank(self) -> int:
+        ...
+
+
+class Collective(ABC):
+    def __init__(self, instantiate_group: bool = False, **group_kwargs: Any) -> None:
+        self._group_kwargs = group_kwargs
+        self._group: Optional[CollectibleGroup] = None
+        if instantiate_group:
+            self.create_group()
+
+    def create_group(self, **kwargs: Any) -> Self:  # type: ignore[valid-type]
```

Comment:
TODO: we need to reorder the properties, methods, abstractmethods for all classes eventually